picoquic/frames.c

```diff
@@ -975,84 +1069,139 @@ int picoquic_prepare_stream_frame(picoquic_cnx_t* cnx, picoquic_stream_head* str
-                if (l_len == 0 || (l_len == space && length > 0)) {
-                    /* Will not try a silly encoding */
-                    *consumed = 0;
-                    ret = PICOQUIC_ERROR_FRAME_BUFFER_TOO_SMALL;
-                } else if (length + l_len > space) {
-                    /* try a shorter packet */
-                    length = space - l_len;
-                    l_len = picoquic_varint_encode(bytes + byte_index, space,
-                        (uint64_t)length);
-                    byte_index += l_len;
-                } else {
-                    /* This is good */
-                    byte_index += l_len;
+                if ((cnx->callback_fn)(cnx, stream->stream_id, (uint8_t*)&stream_data_context, allowed_space, picoquic_callback_prepare_to_send, cnx->callback_ctx) != 0) {
+                    /* something went wrong */
+                    ret = picoquic_connection_error(cnx, PICOQUIC_TRANSPORT_INTERNAL_ERROR, 0);
+                }
+                else {
+                    byte_index = stream_data_context.byte_index + stream_data_context.length;
+                    stream->sent_offset += stream_data_context.length;
+                    cnx->data_sent += stream_data_context.length;
+                    *consumed = byte_index;
+
+                    if (stream_data_context.is_fin) {
+                        stream->is_active = 0;
+                        stream->fin_requested = 1;
+                        stream->fin_sent = 1;
+                        picoquic_update_max_stream_ID_local(cnx, stream);
+                    }
+                    else {
+                        stream->is_active = stream_data_context.is_still_active;
+                    }
                 }
             }
+            else {
+                size_t start_index = 0;
 
-            if (ret == 0 && length > 0 && stream->send_queue != NULL && stream->send_queue->bytes != NULL) {
-                memcpy(&bytes[byte_index], stream->send_queue->bytes + stream->send_queue->offset, length);
-                byte_index += length;
+                if (stream->send_queue == NULL) {
+                    length = 0;
+                }
+                else {
+                    length = (size_t)(stream->send_queue->length - stream->send_queue->offset);
+                }
 
-                stream->send_queue->offset += length;
-                if (stream->send_queue->offset >= stream->send_queue->length) {
-                    picoquic_stream_data* next = stream->send_queue->next_stream_data;
-                    free(stream->send_queue->bytes);
-                    free(stream->send_queue);
-                    stream->send_queue = next;
+                if (length >= allowed_space) {
+                    length = allowed_space;
                 }
 
-                stream->sent_offset += length;
-                cnx->data_sent += length;
-            }
-            *consumed = byte_index;
+                if (length < byte_space) {
```

Comment:
This looks very much like the code above.  And it is a lot of complex logic here.  Refactor into a single function?

Reply:
Yes. There is one slight difference, but it can be accommodated.